analyzer/topology_replication_endpoint.go

```diff
@@ -0,0 +1,302 @@
+
+	// will trigger shttp.WSSpeakerEventHandler, so OnConnected
+	p.wsclient.AddEventHandler(p)
+
+	p.wsclient.Connect()
+}
+
+func (p *TopologyReplicatorPeer) disconnect() {
+	if p.wsclient != nil {
+		p.wsclient.Disconnect()
+	}
+}
+
+func (t *TopologyReplicationEndpoint) addPeer(url *url.URL, auth *shttp.AuthenticationOpts, g *graph.Graph) {
+	peer := &TopologyReplicatorPeer{
+		URL:         url,
+		Graph:       g,
+		AuthOptions: auth,
+	}
+
+	t.peers = append(t.peers, peer)
+}
+
+// ConnectPeers starts a goroutine connecting all the peers.
+func (t *TopologyReplicationEndpoint) ConnectPeers() {
+	for _, peer := range t.peers {
+		t.wg.Add(1)
+		go peer.connect(&t.wg)
+	}
+}
+
+// DisconnectPeers disconnects all the peers and wait until all disconnected.
+func (t *TopologyReplicationEndpoint) DisconnectPeers() {
+	for _, peer := range t.peers {
+		peer.disconnect()
+	}
+	t.wg.Wait()
+}
+
+// OnWSJSONMessage is triggered by message coming from an other peer.
+func (t *TopologyReplicationEndpoint) OnWSJSONMessage(c shttp.WSSpeaker, msg *shttp.WSJSONMessage) {
+	msgType, obj, err := graph.UnmarshalWSMessage(msg)
+	if err != nil {
+		logging.GetLogger().Errorf("Graph: Unable to parse the event %v: %s", msg, err.Error())
+		return
+	}
+
+	t.Graph.Lock()
+	defer t.Graph.Unlock()
+
+	t.replicateMsg.Store(false)
+	defer t.replicateMsg.Store(true)
+
+	t.cached.SetMode(graph.CacheOnlyMode)
+	defer t.cached.SetMode(graph.DefaultMode)
+
+	switch msgType {
+	case graph.SyncRequestMsgType:
+		reply := msg.Reply(t.Graph, graph.SyncReplyMsgType, http.StatusOK)
+		c.SendMessage(reply)
+	case graph.HostGraphDeletedMsgType:
+		logging.GetLogger().Debugf("Got %s message for host %s", graph.HostGraphDeletedMsgType, obj.(string))
+		t.Graph.DelHostGraph(obj.(string))
+	case graph.SyncMsgType, graph.SyncReplyMsgType:
+		r := obj.(*graph.SyncMsg)
+		for _, n := range r.Nodes {
+			if t.Graph.GetNode(n.ID) == nil {
+				t.Graph.NodeAdded(n)
+			}
+		}
+		for _, e := range r.Edges {
+			if t.Graph.GetEdge(e.ID) == nil {
+				t.Graph.EdgeAdded(e)
+			}
+		}
+	case graph.NodeUpdatedMsgType:
+		t.Graph.NodeUpdated(obj.(*graph.Node))
+	case graph.NodeDeletedMsgType:
+		t.Graph.NodeDeleted(obj.(*graph.Node))
+	case graph.NodeAddedMsgType:
+		t.Graph.NodeAdded(obj.(*graph.Node))
+	case graph.EdgeUpdatedMsgType:
+		t.Graph.EdgeUpdated(obj.(*graph.Edge))
+	case graph.EdgeDeletedMsgType:
+		t.Graph.EdgeDeleted(obj.(*graph.Edge))
+	case graph.EdgeAddedMsgType:
+		t.Graph.EdgeAdded(obj.(*graph.Edge))
+	}
+}
+
+// SendToPeers sends the message to all the peers
+func (t *TopologyReplicationEndpoint) notifyPeers(msg *shttp.WSJSONMessage) {
+	for _, p := range t.peers {
+		if p.wsclient != nil {
+			p.wsclient.SendMessage(msg)
+		}
+	}
+}
+
+// OnNodeUpdated graph node updated event. Implements the GraphEventListener interface.
+func (t *TopologyReplicationEndpoint) OnNodeUpdated(n *graph.Node) {
+	msg := shttp.NewWSJSONMessage(graph.Namespace, graph.NodeUpdatedMsgType, n)
```

Comment:
msg could be moved inside the if statement

Reply:
done